src/System.Private.CoreLib/shared/System/IO/MemoryStream.cs

```diff
@@ -547,6 +547,51 @@ public override Task CopyToAsync(Stream destination, int bufferSize, Cancellatio
             }
         }
 
+        public override void CopyTo(Buffers.ReadOnlySpanAction<byte, object?> callback, object? state, int bufferSize)
+        {
+            StreamHelpers.ValidateCopyToArgs(this, callback, bufferSize);
+
+            // If we have been inherited into a subclass, the following implementation could be incorrect
+            // since it does not call through to Read() which a subclass might have overridden.
+            // To be safe we will only use this implementation in cases where we know it is safe to do so,
+            // and delegate to our base class (which will call into Read) when we are not sure.
+            if (GetType() != typeof(MemoryStream))
+            {
+                base.CopyTo(callback, state, bufferSize);
+                return;
+            }
+
+            // Retrieve a span until the end of the MemoryStream.
+            ReadOnlySpan<byte> span = new ReadOnlySpan<byte>(_buffer, _position, _length - _position);
+            _position = _length;
+
+            // Invoke the callback, using our internal span and avoiding any
+            // intermediary allocations.
+            callback(span, state);
+        }
+
+        public override ValueTask CopyToAsync(Func<ReadOnlyMemory<byte>, object?, CancellationToken, ValueTask> callback, object? state, int bufferSize, CancellationToken cancellationToken)
```

Comment:
For now we should change the return type of CopyToAsync back to Task.  We can revisit that, but I'd like to discuss that in API review if we want to.

The delegate returning ValueTask has little downside, as there are very few consumers of that delegate invocation (just someone overriding CopyToAsync) and thus usability of ValueTask isn't really an issue, and we can better map to the perf-focused WriteAsync override that returns a ValueTask.  But for the overall CopyToAsync method returning ValueTask, it is more common to want to do things like have multiple consumers of the returned task, and using a ValueTask here will potentially hamper that usability.  It's also the non-generic, which means we'll already be able to use a cached Task for any synchronous completions, so the only time there will be an extra allocation is for the async completion.  And for the base implementation, we're already incurring a Stream allocation, anyway.

Reply:
Rolled back. Thanks for explaining the rationales 👍

Reply:
Good night 🌃